scipy/special/_basic.py

```diff
@@ -2951,6 +2951,32 @@ def corr(k, r): return np.power(k, -r / k) / gamma(r / k + 1) * r
     return result
 
 
+def _is_subdtype(dtype, dtypes):
+    """
+    Shorthand for calculating whether dtype is subtype of some dtypes.
+
+    Also allows specifying a list instead of just a single dtype.
+
+    Additionaly, the most important supertypes from
+        https://numpy.org/doc/stable/reference/arrays.scalars.html
+    can optionally be specified using abbreviations as follows:
+        "i": np.integer
+        "f": np.floating
+        "c": np.complexfloating
+        "n": np.number (contains the other three)
+    """
+    dtypes = dtypes if isinstance(dtypes, list) else [dtypes]
+    # map single character abbreviations, if they are in dtypes
+    mapping = {
+        "i": np.integer,
+        "f": np.floating,
+        "c": np.complexfloating,
+        "n": np.number
+    }
+    dtypes = [mapping.get(x, x) for x in dtypes]
+    return any(np.issubdtype(dtype, dt) for dt in dtypes)
```

Comment:
this looks good but we'll be able to replace it with just `xp.isdtype` when array API standard support is added

Reply:
You might want to check out https://github.com/scipy/scipy/pull/19988 to see what I'm doing with this function in the end. :)

Not sure if that's part of the design for `xp.isdtype` or whether it should be, but it's certainly a useful thing to have, rather than other extremely verbose variants of the same thing.

(try writing out `_is_subdtype(type(n), ["i", "f", "c"])` "by hand" to see what I mean)

Reply:
yep, in the standard that would be spelled `xp.isdtype(n.dtype, 'numeric')` https://data-apis.org/array-api/draft/API_specification/generated/array_api.isdtype.html.

Reply:
Bad example on my part then. Consider:
```python
if _is_subdtype(type(n), ["f", "c"]):
if _is_subdtype(type(n), ["i", "f", "c", type(None)]):
```
or
```python
# factorial allows floats also for extend="zero"
types_requiring_complex = "c" if fname == "factorial" else ["f", "c"]
if _is_subdtype(type(n), types_requiring_complex):
```

What I'm opposed to is writing out the full names like `"real floating"` all the time. For the kind of code that's necessary for the factorial functions, this leads to an unreadable mess. I think single letters are clear (enough) in context and the gained improvement in legibility is so substantial that I consider it a necessity.

However, that helper doesn't have to leak beyond `special/_basic.py` and I'm also perfectly happy to have `_is_subdtype` dispatch to `xp.isdtype` under the hood.